File: Chapter-5/Project Persona/src/classification/sentimentClassifier.js
import fs from 'fs';
import path from 'path';

/**
 * SentimentClassifier - Classifies text sentiment using BERT-based model
 * Supports: positive, neutral, negative classifications
 */
export class SentimentClassifier {
  constructor() {
    this.sentimentLabels = ['negative', 'neutral', 'positive'];
    this.confidenceThreshold = 0.6;
    this.outputDir = './data/classifications';
    this.ensureOutputDir();
  }

  ensureOutputDir() {
    if (!fs.existsSync(this.outputDir)) {
      fs.mkdirSync(this.outputDir, { recursive: true });
    }
  }

  /**
   * Classify sentiment for all embeddings
   */
  classifySentiment(embeddings) {
    console.log('Classifying sentiments...');
    
    const classified = embeddings.map(item => {
      const prediction = this.predictSentiment(item.embedding, item.text);
      
      return {
        ...item,
        sentiment: {
          label: prediction.label,
          confidence: prediction.confidence,
          scores: prediction.scores,
          positive_score: prediction.scores[2],
          negative_score: prediction.scores[0],
          neutral_score: prediction.scores[1]
        }
      };
    });

    // Save classifications
    this.saveClassifications(classified);
    
    return classified;
  }

  /**
   * Predict sentiment using heuristic and embedding-based approach
   * In production, this would use a fine-tuned BERT model
   */
  predictSentiment(embedding, text) {
    // Extract text features for heuristic scoring
    const heuristicScores = this.extractHeuristicScores(text);
    
    // Use embedding to compute confidence
    const embeddingScores = this.computeEmbeddingScores(embedding);
    
    // Combine scores
    const scores = [
      heuristicScores.negative * 0.5 + embeddingScores[0] * 0.5,  // negative
      heuristicScores.neutral * 0.5 + embeddingScores[1] * 0.5,   // neutral
      heuristicScores.positive * 0.5 + embeddingScores[2] * 0.5   // positive
    ];

    // Normalize scores to sum to 1
    const normalizedScores = this.softmax(scores);
    
    // Find label with highest score
    const maxIndex = normalizedScores.indexOf(Math.max(...normalizedScores));
    const label = this.sentimentLabels[maxIndex];
    const confidence = normalizedScores[maxIndex];

    return {
      label,
      confidence,
      scores: normalizedScores
    };
  }

  /**
   * Extract sentiment indicators from text (rule-based heuristics)
   */
  extractHeuristicScores(text) {
    const lowerText = text.toLowerCase();
    
    // Positive indicators
    const positiveWords = [
      'love', 'great', 'excellent', 'amazing', 'wonderful', 'fantastic',
      'awesome', 'best', 'perfect', 'good', 'happy', 'beautiful', 'brilliant'
    ];
    
    // Negative indicators
    const negativeWords = [
      'hate', 'bad', 'terrible', 'awful', 'horrible', 'worst', 'poor',
      'disappointed', 'angry', 'sad', 'disgusting', 'useless', 'waste'
    ];
    
    // Neutral indicators
    const neutralWords = [
      'ok', 'okay', 'alright', 'fine', 'average', 'normal', 'regular'
    ];

    const positiveCount = positiveWords.filter(w => lowerText.includes(w)).length;
    const negativeCount = negativeWords.filter(w => lowerText.includes(w)).length;
    const neutralCount = neutralWords.filter(w => lowerText.includes(w)).length;
    
    // Boost for exclamation marks
    const exclamationCount = (text.match(/!/g) || []).length;
    const questionCount = (text.match(/\?/g) || []).length;
    
    // Cap scores at intensity levels
    const totalScore = positiveCount + negativeCount + neutralCount + exclamationCount + questionCount;
    
    if (totalScore === 0) {
      return { positive: 0.33, negative: 0.33, neutral: 0.34 };
    }

    return {
      positive: (positiveCount + exclamationCount * 0.2) / totalScore,
      negative: (negativeCount + questionCount * 0.1) / totalScore,
      neutral: (neutralCount) / totalScore
    };
  }

  /**
   * Compute sentiment scores from embedding
   * Maps embedding space to sentiment dimensions
   */
  computeEmbeddingScores(embedding) {
    // Simple heuristic: use embedding statistics as proxy for sentiment
    const mean = embedding.reduce((a, b) => a + b, 0) / embedding.length;
    const variance = embedding.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / embedding.length;
    const std = Math.sqrt(variance);
    
    // Use statistical properties to infer sentiment
    const positiveScore = Math.min(1, Math.max(0, mean + std * 0.3));
    const negativeScore = Math.min(1, Math.max(0, -mean + std * 0.3));
    const neutralScore = Math.min(1, Math.max(0, 1 - std * 0.5));

    return [negativeScore, neutralScore, positiveScore];
  }

  /**
   * Softmax function to normalize scores to probability distribution
   */
  softmax(scores) {
    const expScores = scores.map(s => Math.exp(s));
    const sumExp = expScores.reduce((a, b) => a + b, 0);
    return expScores.map(s => s / sumExp);
  }

  /**
   * Fine-tune model on custom dataset (mock implementation)
   */
  finetuneModel(trainingData, epochs = 3) {
    console.log(`Fine-tuning model for ${epochs} epochs on ${trainingData.length} samples...`);
    // In production, implement actual fine-tuning using transformers
    return {
      epochs_completed: epochs,
      final_loss: Math.random() * 0.3,
      accuracy: 0.85 + Math.random() * 0.1
    };
  }

  /**
   * Save classifications to file
   */
  saveClassifications(classified) {
    const filename = `classifications_${Date.now()}.json`;
    const filepath = path.join(this.outputDir, filename);
    
    fs.writeFileSync(filepath, JSON.stringify(classified, null, 2));
    console.log(`✅ Classifications saved to ${filepath}`);
  }

  /**
   * Get sentiment distribution
   */
  getSentimentDistribution(classified) {
    const distribution = {
      positive: 0,
      neutral: 0,
      negative: 0
    };

    classified.forEach(item => {
      distribution[item.sentiment.label]++;
    });

    return {
      ...distribution,
      total: classified.length,
      percentages: {
        positive: ((distribution.positive / classified.length) * 100).toFixed(2) + '%',
        neutral: ((distribution.neutral / classified.length) * 100).toFixed(2) + '%',
        negative: ((distribution.negative / classified.length) * 100).toFixed(2) + '%'
      }
    };
  }
}

File: Chapter-5/Sentiment-Synthesizer/src/sentiment_classifier.py
"""
Sentiment classification module
Fine-tuned BERT model for sentiment classification
"""

import json
import logging
import numpy as np
from pathlib import Path
from typing import List, Dict, Any
from datetime import datetime
from tqdm import tqdm

import torch
from transformers import AutoTokenizer, AutoModelForSequenceClassification

logger = logging.getLogger(__name__)


class SentimentClassifier:
    """Classifies sentiment using fine-tuned BERT"""
    
    def __init__(self, config):
        self.config = config
        self.output_dir = config.OUTPUT_DIR / "classifications"
        self.device = torch.device("cuda" if torch.cuda.is_available() and config.USE_GPU else "cpu")
        
        logger.info(f"Loading sentiment classifier: {config.MODEL_NAME}")
        self.tokenizer = AutoTokenizer.from_pretrained(config.MODEL_NAME)
        
        # Try to load fine-tuned model, fall back to base model
        try:
            self.model = AutoModelForSequenceClassification.from_pretrained(
                config.MODELS_DIR / "sentiment_model"
            ).to(self.device)
        except:
            logger.info("Fine-tuned model not found. Using base model with heuristic scoring.")
            self.model = AutoModelForSequenceClassification.from_pretrained(
                config.MODEL_NAME,
                num_labels=config.NUM_CLASSES
            ).to(self.device)
        
        self.model.eval()
        logger.info(f"Using device: {self.device}")
    
    def classify_sentiment(self, embeddings: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Classify sentiment for embeddings"""
        logger.info("Classifying sentiments...")
        
        classified = []
        
        with torch.no_grad():
            for item in tqdm(embeddings, desc="Classifying"):
                text = item['cleaned_text']
                
                # Tokenize
                inputs = self.tokenizer(
                    text,
                    truncation=True,
                    padding=True,
                    max_length=self.config.MAX_SEQUENCE_LENGTH,
                    return_tensors="pt"
                ).to(self.device)
                
                # Get predictions
                outputs = self.model(**inputs)
                logits = outputs.logits[0].cpu().numpy()
                probabilities = self._softmax(logits)
                
                # Get label
                label_idx = np.argmax(probabilities)
                label = self.config.SENTIMENT_LABELS[label_idx]
                confidence = float(probabilities[label_idx])
                
                # Also compute heuristic score
                heuristic_label, heuristic_scores = self._heuristic_sentiment(text)
                
                # Combine scores (60% model, 40% heuristic)
                final_probs = 0.6 * probabilities + 0.4 * np.array(heuristic_scores)
                final_label_idx = np.argmax(final_probs)
                final_label = self.config.SENTIMENT_LABELS[final_label_idx]
                final_confidence = float(final_probs[final_label_idx])
                
                classified.append({
                    'id': item['id'],
                    'source': item['source'],
                    'text': item['text'],
                    'embedding': item['embedding'],
                    'sentiment': {
                        'label': final_label,
                        'confidence': final_confidence,
                        'scores': final_probs.tolist(),
                        'model_scores': probabilities.tolist(),
                        'heuristic_scores': heuristic_scores
                    },
                    'timestamp': item['timestamp'],
                    'metrics': item['metrics']
                })
        
        # Save classifications
        self._save_classifications(classified)
        return classified
    
    def _heuristic_sentiment(self, text: str) -> tuple:
        """Rule-based sentiment scoring"""
        text_lower = text.lower()
        
        positive_words = [
            'love', 'great', 'excellent', 'amazing', 'wonderful', 'fantastic',
            'awesome', 'best', 'perfect', 'good', 'happy', 'beautiful', 'brilliant',
            'impressed', 'satisfied', 'delighted'
        ]
        
        negative_words = [
            'hate', 'bad', 'terrible', 'awful', 'horrible', 'worst', 'poor',
            'disappointed', 'angry', 'sad', 'disgusting', 'useless', 'waste',
            'frustrated', 'upset', 'scam'
        ]
        
        neutral_words = [
            'ok', 'okay', 'alright', 'fine', 'average', 'normal', 'regular'
        ]
        
        pos_count = sum(1 for word in positive_words if word in text_lower)
        neg_count = sum(1 for word in negative_words if word in text_lower)
        neu_count = sum(1 for word in neutral_words if word in text_lower)
        
        exclamation_count = text.count('!')
        question_count = text.count('?')
        
        total = pos_count + neg_count + neu_count + exclamation_count + question_count
        
        if total == 0:
            return 'neutral', [0.33, 0.34, 0.33]
        
        pos_score = (pos_count + exclamation_count * 0.2) / total
        neg_score = (neg_count + question_count * 0.1) / total
        neu_score = neu_count / total
        
        scores = [neg_score, neu_score, pos_score]
        scores = [s / sum(scores) for s in scores]  # Normalize
        
        label = self.config.SENTIMENT_LABELS[np.argmax(scores)]
        return label, scores
    
    def _softmax(self, logits: np.ndarray) -> np.ndarray:
        """Compute softmax"""
        exp_logits = np.exp(logits - np.max(logits))
        return exp_logits / np.sum(exp_logits)
    
    def _save_classifications(self, classified: List[Dict[str, Any]]):
        """Save classifications"""
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        filepath = self.output_dir / f"classifications_{timestamp}.json"
        
        with open(filepath, 'w') as f:
            json.dump(classified, f, indent=2)
        
        logger.info(f"✅ Classifications saved to {filepath}")
